agent_build_refactored/tools/runner.py

```diff
@@ -0,0 +1,910 @@
+            ]
+        )
+
+        env_options = []
+        for env_var_name, env_var_val in env_variables_to_pass.items():
+            env_options.extend(["-e", f"{env_var_name}={env_var_val}"])
+
+        check_call_with_log(
+            [
+                "docker",
+                "run",
+                "-i",
+                "--name",
+                self._step_container_name,
+                "--workdir",
+                str(final_isolated_source_root),
+                *mount_options,
+                *env_options,
+                self._base_docker_image.name,
+                *command_args,
+            ]
+        )
+
+    def run(self, work_dir: pl.Path):
+        """
+        Run the step. Based on its initial data, it will be performed in docker or locally, on the current system.
+        """
+
+        output_directory = self.get_output_directory(work_dir)
+        cache_directory = self.get_cache_directory(work_dir)
+        isolated_source_root = self.get_isolated_root(work_dir)
+
+        output_directory.parent.mkdir(parents=True, exist_ok=True)
+        skipped = self._restore_cache(
+            output_directory=output_directory, cache_directory=cache_directory
+        )
+        if not skipped:
+            logging.info(f"Run step {self.name}.")
+            for step in self.required_steps.values():
+                step.run(work_dir=work_dir)
+
+            if self._base_step:
+                self._base_step.run(work_dir=work_dir)
+
+            self._remove_output_directory(output_directory=output_directory)
+            output_directory.mkdir(parents=True, exist_ok=True)
+
+            # Create directory to store only tracked files.
+            if isolated_source_root.exists():
+                shutil.rmtree(isolated_source_root)
+            isolated_source_root.mkdir(parents=True)
+
+            # Copy all tracked files into a new isolated directory.
+            for file_path in self._tracked_files:
+                dest_path = isolated_source_root / file_path.parent.relative_to(
+                    SOURCE_ROOT
+                )
+                dest_path.mkdir(parents=True, exist_ok=True)
+                shutil.copy2(file_path, dest_path)
+
+            all_env_variables = self._get_all_environment_variables(work_dir=work_dir)
+            env_variables_str = "\n    ".join(
+                f"{n}='{v}'" for n, v in all_env_variables.items()
+            )
+            logging.info(
+                f"Start step: {self.id}\n"
+                f"Passed env. variables:\n    {env_variables_str}\n"
+            )
+            try:
+                self._run_script(work_dir=work_dir)
+            except Exception:
+                globs = [str(g) for g in self.tracked_files_globs]
+                logging.exception(
+                    f"'{type(self).__name__}' has failed. "
+                    "HINT: Make sure that you have specified all files. "
+                    f"For now, tracked files are: {globs}."
```

Comment:
Perhaps for troubleshooting it may also make sense to include what those globs expand to?

Reply:
I though about it, but was worried that it might be too verbose if there are many files.